### PR TITLE
OpenCL: some optimizations for Intel Gen Graphics

### PR DESCRIPTION
**Pull request by gongzg**
This PR added some new convolution kernels into the spatial convolution engine and also optimized some other layers such as LRN layer/batchnorm layer/softmax layer and softmax loss layer. We also create an intel isaac project to optimize those blas functions via OpenCL. Please use github.com/intel/isaac with Intel platform.

### CODE REVIEW

**Comment by naibaf7** [comment]
Ok cool, I will review this in the next few days :)

**Comment by atlury** [comment]
@naibaf7 @gongzg Does this mean after this PR merge, we will need to use isaac with say clbas to get better performance?

**Comment by gongzg** [comment]
@atlury Exactly. clBLAS is much slower than the isaac.

**Comment by naibaf7** [comment]
@gongzg The point is also that we have windows support now, but it's not so easy to get all the external libraries working there (yet). Also the Intel spatial kernels are not fully working yet, so that's also something I'm currently investigating. :)

**Comment by gongzg** [comment]
@naibaf7 oh. I have not been aware of the windows support so far. The Intel spatial convolution should only work on Linux now. May be we can start to enable windows support in next step. Thanks.

**Comment by naibaf7** [comment]
@gongzg I thought I wrote you an email about it 2 weeks ago... :)
However yes, that would be cool to get working as well.

**Comment by gongzg** [comment]
@naibaf7 Don't know why I missed your previous email. Too busy these days :).  But it's cool to support both Linux and Windows. Will work on that latter.

**Comment by atlury** [comment]
@naibaf7 @gongzg I have tested these merge commits on an Intel tv stick, webcam using Intel Spatial kernels and using 19-layer model vgg model. I am able to get real time classification. 

Amazing feat!! And all under 3.5 watts. This blows out nvidia tx1. Congrats to you guys.

Next would be to test object detection. (one Warning I encountered is "reading dangerously large protocol message, if the message turns out to be larger, parsing will be halted for security reasons" nevertheless this should be advertised more)

curtain

**Comment by naibaf7** [comment]
@atlury Thanks a lot :) It has been a lot of work to get here by @gongzg, various community contributions and me.

**Comment by naibaf7** [comment]
@gongzg Merged. There have been a few minor issues but I think it should be good now.
